datahub-upgrade/src/main/java/com/linkedin/datahub/upgrade/restorebackup/RestoreStorageStep.java

```diff
@@ -0,0 +1,139 @@
+package com.linkedin.datahub.upgrade.restorebackup;
+
+import com.google.common.collect.ImmutableList;
+import com.linkedin.common.AuditStamp;
+import com.linkedin.common.urn.Urn;
+import com.linkedin.data.template.RecordTemplate;
+import com.linkedin.datahub.upgrade.UpgradeContext;
+import com.linkedin.datahub.upgrade.UpgradeStep;
+import com.linkedin.datahub.upgrade.UpgradeStepResult;
+import com.linkedin.datahub.upgrade.impl.DefaultUpgradeStepResult;
+import com.linkedin.datahub.upgrade.restorebackup.backupreader.EbeanAspectBackupIterator;
+import com.linkedin.datahub.upgrade.restorebackup.backupreader.BackupReader;
+import com.linkedin.datahub.upgrade.restorebackup.backupreader.LocalParquetReader;
+import com.linkedin.metadata.entity.EntityService;
+import com.linkedin.metadata.entity.ebean.EbeanAspectV2;
+import com.linkedin.metadata.entity.ebean.EbeanUtils;
+import com.linkedin.metadata.models.EntitySpec;
+import com.linkedin.metadata.models.registry.EntityRegistry;
+import java.net.URISyntaxException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+
+public class RestoreStorageStep implements UpgradeStep {
+
+  private static final int REPORT_BATCH_SIZE = 1000;
+
+  private final EntityService _entityService;
+  private final EntityRegistry _entityRegistry;
+  private final Map<String, BackupReader> _backupReaders;
+
+  public RestoreStorageStep(final EntityService entityService, final EntityRegistry entityRegistry) {
+    _entityService = entityService;
+    _entityRegistry = entityRegistry;
+    _backupReaders = ImmutableList.of(new LocalParquetReader())
+        .stream()
+        .collect(Collectors.toMap(BackupReader::getName, Function.identity()));
+  }
+
+  @Override
+  public String id() {
+    return "RestoreStorageStep";
+  }
+
+  @Override
+  public int retryCount() {
+    return 0;
+  }
+
+  @Override
+  public Function<UpgradeContext, UpgradeStepResult> executable() {
+    return (context) -> {
+
+      context.report().addLine("Starting backup restore...");
+      int numRows = 0;
+      Optional<String> backupReaderName = context.parsedArgs().get("BACKUP_READER");
+      if (!backupReaderName.isPresent() || !_backupReaders.containsKey(backupReaderName.get())) {
+        context.report().addLine("BACKUP_READER is not set or is not valid");
+        return new DefaultUpgradeStepResult(id(), UpgradeStepResult.Result.FAILED);
+      }
+
+      EbeanAspectBackupIterator iterator = _backupReaders.get(backupReaderName.get()).getBackupIterator(context);
+      if (iterator == null) {
```

Comment:
It wasn't immediately obvious that the iterator could be null. Should the iterator have some method on it for this?

ie. iterator.hasData()?

Reply:
This is hard because there is no null iterator. I'll make this Optional to make this more explicit.

Reply:
ty